src/logger.ts

```diff
@@ -139,20 +133,47 @@ export class Logger {
 		containerId: string,
 		serviceInfo: { serviceId: number; imageId: number },
 	): Bluebird<void> {
-		return Bluebird.using(this.lock(containerId), () => {
-			return this.attachStream(
-				docker,
-				OutputStream.Stdout,
-				containerId,
-				serviceInfo,
-			).then(() => {
-				return this.attachStream(
-					docker,
-					OutputStream.Stderr,
-					containerId,
-					serviceInfo,
-				);
+		// First detect if we already have an attached log stream
+		// for this container
+		if (containerId in this.containerLogs) {
+			return Bluebird.resolve();
+		}
+
+		return Bluebird.using(this.lock(containerId), async () => {
+			const logs = new ContainerLogs(containerId, docker);
+			this.containerLogs[containerId] = logs;
+			logs.on('error', err => {
+				console.error(`Container log retrieval error: ${err}`);
+				delete this.containerLogs[containerId];
+			});
+			logs.on('log', async logMessage => {
+				this.log(_.merge({}, serviceInfo, logMessage));
```

Comment:
Do you actually want merge behaviour here or are you expecting assign behaviour?